File: internal/config_test.go
package internal

import (
	"os"
	"testing"
)

func TestGetAPIKeyOrError(t *testing.T) {
	const validKey = "a1b2c3d4-e5f6-4a7b-8c9d-0e1f2a3b4c5d"
	restore := setEnv("MAGNET_API_KEY", "")
	defer restore()

	// Missing key
	_, err := GetAPIKeyOrError()
	if err == nil {
		t.Fatal("expected error when MAGNET_API_KEY is empty")
	}
	if err.Error() != "missing Magnet API key. Set MAGNET_API_KEY" {
		t.Errorf("unexpected error: %v", err)
	}

	// Invalid UUID (wrong variant)
	setEnv("MAGNET_API_KEY", "a1b2c3d4-e5f6-4a7b-0c9d-0e1f2a3b4c5d") // 0 = invalid
	_, err = GetAPIKeyOrError()
	if err == nil {
		t.Fatal("expected error for invalid UUID")
	}

	// Valid key
	setEnv("MAGNET_API_KEY", validKey)
	key, err := GetAPIKeyOrError()
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if key != validKey {
		t.Errorf("got key %q", key)
	}

	// Key with whitespace is trimmed
	setEnv("MAGNET_API_KEY", "  "+validKey+"  ")
	key, err = GetAPIKeyOrError()
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if key != validKey {
		t.Errorf("got key %q", key)
	}
}

func TestGetBaseURL(t *testing.T) {
	restore := setEnv("MAGNET_API_URL", "")
	defer restore()

	u := GetBaseURL()
	if u != defaultBaseURL {
		t.Errorf("default base URL: got %q", u)
	}

	setEnv("MAGNET_API_URL", " https://custom.example.com/ ")
	u = GetBaseURL()
	if u != "https://custom.example.com" {
		t.Errorf("custom base URL: got %q", u)
	}
}

func TestAPIKeyRegex(t *testing.T) {
	valid := []string{
		"a1b2c3d4-e5f6-4a7b-8c9d-0e1f2a3b4c5d",
		"A1B2C3D4-E5F6-4A7B-8C9D-0E1F2A3B4C5D",
		"00000000-0000-4000-8000-000000000000",
	}
	for _, s := range valid {
		if !apiKeyRegex.MatchString(s) {
			t.Errorf("expected valid: %q", s)
		}
	}
	invalid := []string{
		"",
		"not-a-uuid",
		"a1b2c3d4-e5f6-4a7b-8c9d-0e1f2a3b4c5d-extra",
		"a1b2c3d4e5f64a7b8c9d0e1f2a3b4c5d", // no dashes
		"a1b2c3d4-e5f6-6a7b-8c9d-0e1f2a3b4c5d", // version 6
	}
	for _, s := range invalid {
		if apiKeyRegex.MatchString(s) {
			t.Errorf("expected invalid: %q", s)
		}
	}
}

func setEnv(key, value string) func() {
	old := os.Getenv(key)
	if value == "" {
		os.Unsetenv(key)
	} else {
		os.Setenv(key, value)
	}
	return func() {
		if old == "" {
			os.Unsetenv(key)
		} else {
			os.Setenv(key, old)
		}
	}
}

File: src/config.test.ts
import { describe, test, expect } from "bun:test";
import {
  getApiKeyOrError,
  getBaseUrl,
  DEFAULT_LIST_LIMIT_VALUE,
} from "./config";

const validKey = "a1b2c3d4-e5f6-4a7b-8c9d-0e1f2a3b4c5d";

describe("config", () => {
  test("getApiKeyOrError returns error when MAGNET_API_KEY empty", () => {
    const orig = process.env.MAGNET_API_KEY;
    delete process.env.MAGNET_API_KEY;
    const r = getApiKeyOrError();
    if (orig !== undefined) process.env.MAGNET_API_KEY = orig;
    expect(r).toBeInstanceOf(Error);
    expect((r as Error).message).toContain("missing");
  });

  test("getApiKeyOrError returns error for invalid UUID", () => {
    const orig = process.env.MAGNET_API_KEY;
    process.env.MAGNET_API_KEY = "not-a-uuid";
    const r = getApiKeyOrError();
    if (orig !== undefined) process.env.MAGNET_API_KEY = orig;
    else delete process.env.MAGNET_API_KEY;
    expect(r).toBeInstanceOf(Error);
  });

  test("getApiKeyOrError returns key for valid UUID", () => {
    const orig = process.env.MAGNET_API_KEY;
    process.env.MAGNET_API_KEY = validKey;
    const r = getApiKeyOrError();
    process.env.MAGNET_API_KEY = orig ?? "";
    expect(r).toBe(validKey);
  });

  test("getBaseUrl returns default when MAGNET_API_URL unset", () => {
    const orig = process.env.MAGNET_API_URL;
    delete process.env.MAGNET_API_URL;
    const u = getBaseUrl();
    if (orig !== undefined) process.env.MAGNET_API_URL = orig;
    expect(u).toBe("https://www.magnet.run");
  });

  test("getBaseUrl trims and strips trailing slash", () => {
    const orig = process.env.MAGNET_API_URL;
    process.env.MAGNET_API_URL = " https://custom.example.com/ ";
    const u = getBaseUrl();
    process.env.MAGNET_API_URL = orig ?? "";
    expect(u).toBe("https://custom.example.com");
  });

  test("DEFAULT_LIST_LIMIT_VALUE is 50", () => {
    expect(DEFAULT_LIST_LIMIT_VALUE).toBe(50);
  });
});
